File: Task/Smith-numbers/Lua/smith-numbers.lua
-- Returns a boolean indicating whether n is prime
function isPrime (n)
    if n < 2 then return false end
    if n < 4 then return true end
    if n % 2 == 0 then return false end
    for d = 3, math.sqrt(n), 2 do
        if n % d == 0 then return false end
    end
    return true
end

-- Returns a table of the prime factors of n
function primeFactors (n)
    local pfacs, divisor = {}, 1
    if n < 1 then return pfacs end
    while not isPrime(n) do
        while not isPrime(divisor) do divisor = divisor + 1 end
        while n % divisor == 0 do
            n = n / divisor
            table.insert(pfacs, divisor)
        end
        divisor = divisor + 1
        if n == 1 then return pfacs end
    end
    table.insert(pfacs, n)
    return pfacs
end

-- Returns the sum of the digits of n
function sumDigits (n)
    local sum, nStr = 0, tostring(n)
    for digit = 1, nStr:len() do
        sum = sum + tonumber(nStr:sub(digit, digit))
    end
    return sum
end

-- Returns a boolean indicating whether n is a Smith number
function isSmith (n)
    if isPrime(n) then return false end
    local sumFacs = 0
    for _, v in ipairs(primeFactors(n)) do
        sumFacs = sumFacs + sumDigits(v)
    end
    return sumFacs == sumDigits(n)
end

-- Main procedure
for n = 1, 10000 do
    if isSmith(n) then io.write(n .. "\t") end
end

File: Task/Smith-numbers/Nim/smith-numbers.nim
import strformat

func primeFactors(n: int): seq[int] =
  result = newSeq[int]()
  var n = n
  var i = 2
  while n mod i == 0:
    result.add(i)
    n = n div i
  i = 3
  while i * i <= n:
    while n mod i == 0:
      result.add(i)
      n = n div i
    inc i, 2
  if n != 1:
    result.add(n)

func sumDigits(n: int): int =
  var n = n
  var sum = 0
  while n > 0:
    inc sum, n mod 10
    n = n div 10
  sum

var cnt = 0
for n in 1..10_000:
  var factors = primeFactors(n)
  if factors.len > 1:
    var sum = sumDigits(n)
    for f in factors:
      dec sum, sumDigits(f)
    if sum == 0:
      stdout.write(&"{n:4}  ")
      inc cnt
    if cnt == 10:
      cnt = 0
      stdout.write("\n")
echo()
